barretenberg/cpp/pil/avm/gadgets/range_check.pil

```diff
@@ -0,0 +1,144 @@
+    // This register has a (more expensive) set of constraint that enables dynamic range check of bit values between 0 and 16 bits
+    pol commit u16_r7;
+
+    // In each of these relations, the u16_r7 register contains the most significant 16 bits of value.
+    pol X_0 = is_lte_u16 * u16_r7;
+    pol X_1 = is_lte_u32 * (u16_r0 + u16_r7 * 2**16);
+    pol X_2 = is_lte_u48 * (u16_r0 + u16_r1 * 2**16 + u16_r7 * 2**32);
+    pol X_3 = is_lte_u64 * (u16_r0 + u16_r1 * 2**16 + u16_r2 * 2**32 + u16_r7 * 2**48);
+    pol X_4 = is_lte_u80 * (u16_r0 + u16_r1 * 2**16 + u16_r2 * 2**32 + u16_r3 * 2**48 + u16_r7 * 2**64);
+    pol X_5 = is_lte_u96 * (u16_r0 + u16_r1 * 2**16 + u16_r2 * 2**32 + u16_r3 * 2**48 + u16_r4 * 2**64 + u16_r7 * 2**80);
+    pol X_6 = is_lte_u112 * (u16_r0 + u16_r1 * 2**16 + u16_r2 * 2**32 + u16_r3 * 2**48 + u16_r4 * 2**64 + u16_r5 * 2**80 + u16_r7 * 2**96);
+    pol X_7 = is_lte_u128 * (u16_r0 + u16_r1 * 2**16 + u16_r2 * 2**32 + u16_r3 * 2**48 + u16_r4 * 2**64 + u16_r5 * 2**80 + u16_r6 * 2**96 + u16_r7 * 2**112);
+
+    // Since the is_lte_x are mutually exclusive, only one of the Xs will be non-zero
+    pol RESULT = X_0 + X_1 + X_2 + X_3 + X_4 + X_5 + X_6 + X_7;
+
+    #[CHECK_RECOMPOSITION]
+    sel_rng_chk * (RESULT - value) = 0;
+
+    // The number of bits that form the dynamic range check is depending on the claimed lte value and the witness rng_chk_bits
+    // claimed is_lte_x | dyn_rng_chk_bits
+    // -----------------|-----------------
+    //    is_lte_u16    |  rng_chk_bits
+    //    is_lte_u32    |  rng_chk_bits - 16
+    //    is_lte_u48    |  rng_chk_bits - 32
+    //    is_lte_u64    |  rng_chk_bits - 48
+    //    is_lte_u80    |  rng_chk_bits - 64
+    //    is_lte_u96    |  rng_chk_bits - 80
+    //    is_lte_u112   |  rng_chk_bits - 96
+    //    is_lte_u128   |  rng_chk_bits - 112
+
+    // The number of bits that need to be dynamically range checked.
+    pol commit dyn_rng_chk_bits;
+    // Valid values for dyn_rng_chk_bits are in the range [0, 16]
+    dyn_rng_chk_bits - (rng_chk_bits - (is_lte_u32 * 16) - (is_lte_u48 * 32) - (is_lte_u64 * 48) - (is_lte_u80 * 64) - (is_lte_u96 * 80) - (is_lte_u112 * 96) - (is_lte_u128 * 112)) = 0;
+
+    // To perform the dynamic range check we also need the value of 2^dyn_rng_chk_bits
+    pol commit dyn_rng_chk_pow_2;
+
+    // This lookup does 2 things (1) Indirectly range checks dyn_rng_chk_bits to not have underflowed and (2) Simplified calculation of 2^dyn_rng_chk_bits
+    #[LOOKUP_RNG_CHK_POW_2]
+    sel_rng_chk {dyn_rng_chk_bits, dyn_rng_chk_pow_2} in main.sel_rng_8 {main.clk, powers.power_of_2};
+
+
+    // Now we need to perform the dynamic range check itself
+    // We check that u16_r7 < dyn_rng_chk_pow_2 ==> dyn_rng_chk_pow_2 - u16_r7 - 1 > 0
+    pol commit dyn_diff;
+    sel_rng_chk * (dyn_diff - (dyn_rng_chk_pow_2 - u16_r7 - 1)) = 0;
+    // The value of dyn_diff has to be between [0, 2^16)
+    // To check we did not underflow we just range check it
```

Comment:
I think we are missing a constraint that dyn_rng_chk_bits < 16, otherwise the activated is_lte_uXX might not be the right one. If nothing is missing, could you please provide a proof? I have the impression that there might corner cases whenever the value is a power of 2^16
I think the following lookup implies that dyn_rng_chk_bits <= 16, but I do not think it is enough.

Reply:
I believe it is fine for `dyn_rng_chk_pow_2` to have the value of `2^16` and for `dyn_rng_chk_bits = 16`.

This happens if the dynamic register is checking a 0 value, trivially when you range check the value 0 fits within 16 bits. 

This makes `dyn_diff = 2^16 - 0 - 1 = 2^16 -1` , which will pass a 16-bit lookup check

Reply:
When rng_chk_bits is a multiple of 16 denoted N, the prover can choose between is_lte_uN and is_lte_u(N+1).
If he chooses the latter, the dyn_rng_chk_bits == 0. This is ok because, this will imply that u16_r7 == 0.

@IlyasRidhuan It would be great to just add a note that is_lte_uXX might not be unique in this case and explaining that the constraints are still fine.